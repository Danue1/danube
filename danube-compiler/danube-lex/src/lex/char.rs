use crate::{Error, Lex};
use danube_span::Span;
use danube_token::{LiteralKind, Token, TokenKind};

impl<'lex> Lex<'lex> {
    pub fn lex_char(&mut self) -> Result<Token, Error> {
        let start = self.cursor.cursor();

        match self.cursor.next() {
            Some('\\') => {
                let c = match self.cursor.next() {
                    Some('r') => '\r',
                    Some('n') => '\n',
                    Some('t') => '\t',
                    Some(c) => c,
                    None => return Err(Error::Invalid(self.cursor.cursor() - 1)),
                };
                match self.cursor.next() {
                    Some('\'') => {
                        let span = Span::new(start - 1, start + 3);
                        let symbol = self.intern(c.to_string().as_ref());
                        let kind = TokenKind::Literal(symbol, LiteralKind::Char);

                        Ok(Token::new(span, kind))
                    }
                    _ => Err(Error::Invalid(start + 2)),
                }
            }
            Some(c) => {
                let span = Span::new(start - 1, start + 2);
                let symbol = self.intern(c.to_string().as_ref());
                let kind = TokenKind::Literal(symbol, LiteralKind::Char);

                Ok(Token::new(span, kind))
            }
            _ => Err(Error::Invalid(start)),
        }
    }
}
