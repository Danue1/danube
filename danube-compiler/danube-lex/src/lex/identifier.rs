use crate::{Cursor, Error, Lex};
use danube_span::Span;
use danube_token::{Symbol, Token, TokenKind};

impl<'lex> Lex<'lex> {
    pub fn lex_identifier(&mut self) -> Result<Token, Error> {
        let span = lex_identifier_postfix(&mut self.cursor);
        let symbol = Symbol::intern(self.cursor.slice(&span));

        Ok(Token::new(span, TokenKind::Identifier(symbol)))
    }
}

fn lex_identifier_postfix(cursor: &mut Cursor) -> Span {
    let start = cursor.location();
    while let Some('a'..='z' | 'A'..='Z' | '0'..='9' | '_') = cursor.peek() {
        cursor.next();
    }
    let end = cursor.location();

    Span::new(start, end)
}
