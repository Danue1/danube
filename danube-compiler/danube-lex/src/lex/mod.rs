mod char;
mod identifier;
mod number;
mod string;

use crate::{Cursor, Error};
use danube_span::Span;
use danube_token::{Symbol, SymbolContainer, SymbolInterner, Token, TokenKind};

pub struct Lex<'lex> {
    interner: SymbolInterner,
    cursor: Cursor<'lex>,
}

impl<'lex> Lex<'lex> {
    pub fn symbols(self) -> SymbolContainer {
        self.interner.into()
    }
}

impl<'lex> std::iter::Iterator for Lex<'lex> {
    type Item = Result<Token, Error>;

    fn next(&mut self) -> Option<Self::Item> {
        loop {
            match self.cursor.peek() {
                Some(' ' | '\r' | '\n' | '\t') => {
                    self.cursor.next();
                }
                Some('0'..='9') => return Some(self.lex_number()),
                Some('a'..='z' | 'A'..='Z') => return Some(self.lex_identifier()),
                Some(_) => return Some(self.lex()),
                None => return None,
            }
        }
    }
}

macro_rules! symbol {
    ($cursor:expr => $kind:ident) => {{
        let symbol = TokenKind::$kind;
        let end = $cursor.cursor();
        let count = match symbol.count() {
            Some(count) => count,
            None => return Err(Error::UnknownSymbol),
        };
        let span = Span::new(end - count, end);

        Ok(Token::new(span, symbol))
    }};
}

impl<'lex> Lex<'lex> {
    pub fn new(source: &'lex str) -> Self {
        Lex {
            interner: Default::default(),
            cursor: Cursor::new(source),
        }
    }

    #[inline(always)]
    fn intern(&mut self, string: &str) -> Symbol {
        self.interner.intern(string)
    }

    #[inline(always)]
    fn lex(&mut self) -> Result<Token, Error> {
        match self.cursor.next().unwrap() {
            '(' => symbol!(self.cursor => LeftParens),
            ')' => symbol!(self.cursor => RightParens),
            '[' => symbol!(&self.cursor => LeftBracket),
            ']' => symbol!(&self.cursor => RightBracket),
            '{' => symbol!(&self.cursor => LeftBrace),
            '}' => symbol!(&self.cursor => RightBrace),
            '<' => match self.cursor.peek() {
                Some('<') => {
                    self.cursor.next();
                    match self.cursor.peek() {
                        Some('=') => {
                            self.cursor.next();
                            symbol!(self.cursor => LeftChevronLeftChevronEq)
                        }
                        _ => symbol!(self.cursor => LeftChevronLeftChevron),
                    }
                }
                Some('=') => {
                    self.cursor.next();
                    symbol!(self.cursor => LeftChevronEq)
                }
                _ => symbol!(self.cursor => LeftChevron),
            },
            '>' => match self.cursor.peek() {
                Some('>') => {
                    self.cursor.next();
                    match self.cursor.peek() {
                        Some('=') => {
                            self.cursor.next();
                            symbol!(self.cursor => RightChevronRightChevronEq)
                        }
                        _ => symbol!(self.cursor => RightChevronRightChevron),
                    }
                }
                Some('=') => {
                    self.cursor.next();
                    symbol!(self.cursor => RightChevronEq)
                }
                _ => symbol!(self.cursor => RightChevron),
            },
            '#' => symbol!(self.cursor => Hash),
            '.' => match self.cursor.peek() {
                Some('.') => {
                    self.cursor.next();
                    match self.cursor.peek() {
                        Some('=') => {
                            self.cursor.next();
                            symbol!(self.cursor => DotDotEq)
                        }
                        _ => symbol!(self.cursor => DotDot),
                    }
                }
                Some('0'..='9') => self.lex_number_without_integer(),
                _ => symbol!(self.cursor => Dot),
            },
            ',' => symbol!(self.cursor => Comma),
            ':' => match self.cursor.peek() {
                Some(':') => {
                    self.cursor.next();
                    symbol!(self.cursor => ColonColon)
                }
                _ => symbol!(self.cursor => Colon),
            },
            ';' => symbol!(self.cursor => Semicolon),
            '=' => match self.cursor.peek() {
                Some('>') => {
                    self.cursor.next();
                    symbol!(self.cursor => EqRightChevron)
                }
                Some('=') => {
                    self.cursor.next();
                    symbol!(self.cursor => EqEq)
                }
                _ => symbol!(self.cursor => Eq),
            },
            '+' => match self.cursor.peek() {
                Some('=') => {
                    self.cursor.next();
                    symbol!(self.cursor => PlusEq)
                }
                _ => symbol!(self.cursor => Plus),
            },
            '-' => match self.cursor.peek() {
                Some('=') => {
                    self.cursor.next();
                    symbol!(self.cursor => HyphenEq)
                }
                Some('>') => {
                    self.cursor.next();
                    symbol!(self.cursor => HyphenRightChevron)
                }
                _ => symbol!(self.cursor => Hyphen),
            },
            '*' => match self.cursor.peek() {
                Some('*') => {
                    self.cursor.next();
                    match self.cursor.peek() {
                        Some('=') => {
                            self.cursor.next();
                            symbol!(self.cursor => AsteriskAsteriskEq)
                        }
                        _ => symbol!(self.cursor => AsteriskAsterisk),
                    }
                }
                Some('=') => {
                    self.cursor.next();
                    symbol!(self.cursor => AsteriskEq)
                }
                _ => symbol!(self.cursor => Asterisk),
            },
            '/' => match self.cursor.peek() {
                Some('/') => {
                    self.cursor.next();
                    let start = self.cursor.cursor() - 2;
                    let is_document = matches!(self.cursor.peek(), Some('/'));
                    if is_document {
                        self.cursor.next();
                    }
                    while !matches!(self.cursor.peek(), Some('\r' | '\n') | None) {
                        self.cursor.next();
                    }
                    let end = self.cursor.cursor();
                    let span = Span::new(start, end);
                    let symbol = self.intern(self.cursor.slice(&span));

                    Ok(Token::new(span, TokenKind::Comment(symbol)))
                }
                Some('*') => {
                    self.cursor.next();
                    std::todo!();
                }
                Some('=') => {
                    self.cursor.next();
                    symbol!(self.cursor => SlashEq)
                }
                _ => symbol!(self.cursor => Slash),
            },
            '%' => match self.cursor.peek() {
                Some('=') => {
                    self.cursor.next();
                    symbol!(self.cursor => PercentEq)
                }
                _ => symbol!(self.cursor => Percent),
            },
            '!' => match self.cursor.peek() {
                Some('=') => {
                    self.cursor.next();
                    symbol!(self.cursor => ExclamationEq)
                }
                _ => symbol!(self.cursor => Exclamation),
            },
            '?' => symbol!(self.cursor => Question),
            '&' => match self.cursor.peek() {
                Some('&') => {
                    self.cursor.next();
                    match self.cursor.peek() {
                        Some('=') => {
                            self.cursor.next();
                            symbol!(self.cursor => AmpersandAmpersandEq)
                        }
                        _ => symbol!(self.cursor => AmpersandAmpersand),
                    }
                }
                Some('=') => {
                    self.cursor.next();
                    symbol!(self.cursor => AmpersandEq)
                }
                _ => symbol!(self.cursor => Ampersand),
            },
            '|' => match self.cursor.peek() {
                Some('|') => {
                    self.cursor.next();
                    match self.cursor.peek() {
                        Some('=') => {
                            self.cursor.next();
                            symbol!(self.cursor => PipelinePipelineEq)
                        }
                        _ => symbol!(self.cursor => PipelinePipeline),
                    }
                }
                Some('=') => {
                    self.cursor.next();
                    symbol!(self.cursor => PipelineEq)
                }
                _ => symbol!(self.cursor => Pipeline),
            },
            '~' => match self.cursor.peek() {
                Some('=') => {
                    self.cursor.next();
                    symbol!(self.cursor => TildeEq)
                }
                _ => symbol!(self.cursor => Tilde),
            },
            '^' => match self.cursor.peek() {
                Some('=') => {
                    self.cursor.next();
                    symbol!(self.cursor => CaretEq)
                }
                _ => symbol!(self.cursor => Caret),
            },
            '_' => match self.cursor.peek() {
                Some('a'..='z' | 'A'..='Z' | '0'..='9') => self.lex_identifier_with_underscore(),
                _ => symbol!(self.cursor => Underscore),
            },
            '"' => self.lex_string(),
            '\'' => self.lex_char(),
            _ => Err(Error::Invalid(self.cursor.cursor())),
        }
    }
}

#[test]
fn skip_whitespace() {
    assert_eq!(Lex::new(" ").next(), None);
    assert_eq!(Lex::new("\r").next(), None);
    assert_eq!(Lex::new("\n").next(), None);
    assert_eq!(Lex::new("\t").next(), None);
    assert_eq!(Lex::new(" \r\n\t").next(), None);
}

#[test]
fn symbols() {
    macro_rules! assert_symbols {
        ($($symbol:expr => $expect:ident,)+) => {
            $(
                let kind = TokenKind::$expect;
                if let Some(count) = kind.count() {
                    assert_eq!(
                        Lex::new($symbol).next(),
                        Some(Ok(Token::new(Span::new(0, count), kind)))
                    );
                } else {
                    // assert with error message for preventing human-mistakes.
                    std::todo!();
                }
            )+
        };
    }

    assert_symbols! {
        "(" => LeftParens,
        ")" => RightParens,
        "[" => LeftBracket,
        "]" => RightBracket,
        "{" => LeftBrace,
        "}" => RightBrace,
        "<" => LeftChevron,
        ">" => RightChevron,
        "#" => Hash,
        "." => Dot,
        "," => Comma,
        ":" => Colon,
        ";" => Semicolon,
        "=" => Eq,
        "+" => Plus,
        "-" => Hyphen,
        "*" => Asterisk,
        "/" => Slash,
        "%" => Percent,
        "!" => Exclamation,
        "?" => Question,
        "&" => Ampersand,
        "|" => Pipeline,
        "~" => Tilde,
        "^" => Caret,
        "_" => Underscore,

        "->" => HyphenRightChevron,
        "=>" => EqRightChevron,
        ".." => DotDot,
        "::" => ColonColon,
        "==" => EqEq,
        "!=" => ExclamationEq,
        "+=" => PlusEq,
        "-=" => HyphenEq,
        "*=" => AsteriskEq,
        "/=" => SlashEq,
        "%=" => PercentEq,
        "**" => AsteriskAsterisk,
        "&&" => AmpersandAmpersand,
        "||" => PipelinePipeline,
        "&=" => AmpersandEq,
        "|=" => PipelineEq,
        "~=" => TildeEq,
        "^=" => CaretEq,
        "<<" => LeftChevronLeftChevron,
        ">>" => RightChevronRightChevron,
        "<=" => LeftChevronEq,
        ">=" => RightChevronEq,

        "..=" => DotDotEq,
        "**=" => AsteriskAsteriskEq,
        "&&=" => AmpersandAmpersandEq,
        "||=" => PipelinePipelineEq,
        "<<=" => LeftChevronLeftChevronEq,
        ">>=" => RightChevronRightChevronEq,
    };
}

#[test]
fn integers() {
    use danube_token::{LiteralKind, SymbolInterner};

    macro_rules! assert_integers {
        ($($integer:expr),+) => {
            $(
                let mut interner = SymbolInterner::default();

                assert_eq!(
                    Lex::new($integer).next(),
                    Some(Ok(Token::new(
                        Span::new(0, $integer.len()),
                        TokenKind::Literal(interner.intern($integer), LiteralKind::Integer),
                    )))
                );
            )+
        };
    }

    assert_integers!["0", "1", "10", "11"];
}

#[test]
fn floatings() {
    macro_rules! assert_floatings {
        ($($floating:expr),+) => {
            use danube_token::{LiteralKind, SymbolInterner};

            $(
                let mut interner = SymbolInterner::default();

                assert_eq!(
                    Lex::new($floating).next(),
                    Some(Ok(Token::new(
                        Span::new(0, $floating.len()),
                        TokenKind::Literal(interner.intern($floating), LiteralKind::Float),
                    )))
                );
            )+
        };
    }

    assert_floatings![".0", ".1", "0.0", "0.1", "1.23"];
}

#[test]
fn chars() {
    macro_rules! assert_chars {
        ($($char:expr),+) => {
            use danube_token::{LiteralKind, SymbolInterner};

            $(
                let mut interner = SymbolInterner::default();

                assert_eq!(
                    Lex::new($char).next(),
                    Some(Ok(Token::new(
                        Span::new(0, $char.len()),
                        TokenKind::Literal(interner.intern($char), LiteralKind::Char),
                    )))
                );
            )+
        };
    }

    assert_chars![r#"'a'"#, r#"'\r'"#, r#"'\n'"#, r#"'\t'"#];
}

#[test]
fn strings() {
    macro_rules! assert_strings {
        ($($string:expr),+) => {
            use danube_token::{LiteralKind, SymbolInterner};

            $(
                let mut interner = SymbolInterner::default();

                assert_eq!(
                    Lex::new($string).next(),
                    Some(Ok(Token::new(
                        Span::new(0, $string.len()),
                        TokenKind::Literal(interner.intern($string), LiteralKind::String),
                    )))
                );
            )+
        };
    }

    assert_strings![r#""Hello, World!""#, r#""\r\t\n""#];
}

#[test]
fn identifiers() {
    macro_rules! tokens {
        ($($identifier:expr),+ $(,)?) => {{
            let mut interner = SymbolInterner::default();
            let mut vec = Vec::new();
            let mut position = 0;
            $(
                let span = Span::new(position, position + $identifier.len());
                #[allow(unused_assignments)]
                {
                    position = span.end + 1;
                }

                let kind = TokenKind::Identifier(interner.intern($identifier));
                vec.push(Token::new(span, kind));

            )+
            vec
        }};
    }

    let source = "_a _1 _a1 _1a a a_ a1 ab ab1";
    let tokens: Vec<Token> = Lex::new(source).filter_map(|token| token.ok()).collect();

    assert_eq!(
        tokens,
        tokens!["_a", "_1", "_a1", "_1a", "a", "a_", "a1", "ab", "ab1"]
    );
}

#[test]
fn keywords() {
    macro_rules! assert_keywords {
        ($($keyword:expr),+) => {
            $(
                let mut interner = SymbolInterner::default();

                assert_eq!(
                    Lex::new($keyword).next(),
                    Some(Ok(Token::new(
                        Span::new(0, $keyword.len()),
                        TokenKind::Identifier(interner.intern($keyword)),
                    )))
                );
            )+
        };
    }

    assert_keywords![
        "if", "else", "for", "while", "loop", "in", "break", "continue", "match", "return",
        "yield", "where", "const", "let", "mut", "enum", "struct", "fn", "Self", "self", "use",
        "super", "pub", "as", "package", "type", "trait", "impl"
    ];
}

#[test]
fn singleline_comment() {
    macro_rules! assert_comments {
        ($($comment:expr),+) => {
            $(
                let mut interner = SymbolInterner::default();

                assert_eq!(Lex::new($comment).next(), Some(Ok(Token::new(
                    Span::new(0, $comment.len()),
                    TokenKind::Comment(interner.intern($comment))
                ))));
            )+
        };
    }

    assert_comments!["//", "///", "//hello", "///hello"];
}
