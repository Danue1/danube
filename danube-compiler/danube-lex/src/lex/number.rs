use crate::{Cursor, Error, Lex};
use danube_span::Span;
use danube_token::{LiteralKind, Symbol, Token, TokenKind};

impl<'lex> Lex<'lex> {
    pub fn lex_number(&mut self) -> Result<Token, Error> {
        let integer_span = lex_numeric(&mut self.cursor);

        match self.cursor.peek() {
            Some('.') => {
                self.cursor.next();
                match self.cursor.peek() {
                    Some('0'..='9') => {
                        let float_span = integer_span.concat(lex_numeric(&mut self.cursor));
                        let string = self.cursor.slice(&float_span);
                        let symbol = Symbol::intern(string);
                        let kind = TokenKind::Literal(symbol, LiteralKind::Float);

                        Ok(Token::new(float_span, kind))
                    }
                    _ => Err(Error::Invalid(self.cursor.cursor())),
                }
            }
            _ => {
                let string = self.cursor.slice(&integer_span);
                let symbol = Symbol::intern(string);
                let kind = TokenKind::Literal(symbol, LiteralKind::Integer);

                Ok(Token::new(integer_span, kind))
            }
        }
    }

    pub fn lex_number_without_integer(&mut self) -> Result<Token, Error> {
        let float_span = lex_numeric(&mut self.cursor);
        let float_span = Span::new(float_span.start - 1, float_span.end);
        let string = self.cursor.slice(&float_span);
        let symbol = Symbol::intern(string);
        let kind = TokenKind::Literal(symbol, LiteralKind::Float);

        Ok(Token::new(float_span, kind))
    }
}

fn lex_numeric(cursor: &mut Cursor) -> Span {
    let start = cursor.cursor();
    while let Some('0'..='9') = cursor.peek() {
        cursor.next();
    }
    let end = cursor.cursor();

    Span::new(start, end)
}
