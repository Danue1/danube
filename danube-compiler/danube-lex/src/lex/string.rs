use crate::{Error, Lex};
use danube_span::Span;
use danube_token::{LiteralKind, Token, TokenKind};

impl<'lex> Lex<'lex> {
    pub fn lex_string(&mut self) -> Result<Token, Error> {
        let start = self.cursor.cursor();
        let mut string = String::new();

        loop {
            match self.cursor.next() {
                Some('"') => {
                    let span = Span::new(start - 1, self.cursor.cursor());
                    let symbol = self.intern(string.as_ref());
                    let kind = TokenKind::Literal(symbol, LiteralKind::String);

                    return Ok(Token::new(span, kind));
                }
                Some('\\') => {
                    let c = match self.cursor.next() {
                        Some('r') => '\r',
                        Some('n') => '\n',
                        Some('t') => '\t',
                        Some(c) => c,
                        None => return Err(Error::Invalid(self.cursor.cursor() - 1)),
                    };
                    string.push(c);
                }
                Some(c) => {
                    string.push(c);
                }
                None => return Err(Error::Invalid(self.cursor.cursor())),
            }
        }
    }
}
