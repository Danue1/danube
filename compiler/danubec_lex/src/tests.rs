use crate::lex;

#[test]
fn punctuations() {
    let source = "\n\t~`!@#$%^&*-+=|:;,./?\\{}[]()<>";
    let tokens = lex(source);

    insta::assert_debug_snapshot!(tokens);
}

#[test]
fn r#char() {
    let source = "'a' '\\n' '\\'' '한' '\\u{AC00}'";
    let tokens = lex(source);

    insta::assert_debug_snapshot!(tokens);
}

#[test]
fn empty_string() {
    let source = r#""" """""""#;
    let tokens = lex(source);

    insta::assert_debug_snapshot!(tokens);
}

#[test]
fn non_multiline_string() {
    let source = r#""Hello, World!" "a${b}c""#;
    let tokens = lex(source);

    insta::assert_debug_snapshot!(tokens);
}

#[test]
fn multiline_string() {
    let source = r#""""Hello, "World"!""" """a${b}c""""#;
    let tokens = lex(source);

    insta::assert_debug_snapshot!(tokens);
}

#[test]
fn binary() {
    let source = "0b0101 0b_0101 0b__0101 0b01_01 0b01__01 0b0101_ 0b0101__";
    let tokens = lex(source);

    insta::assert_debug_snapshot!(tokens);
}

#[test]
fn octal() {
    let source =
        "0o01234567 0o_01234567 0o__01234567 0o01234_567 0o01234__567 0o01234567_ 0o01234567__";
    let tokens = lex(source);

    insta::assert_debug_snapshot!(tokens);
}

#[test]
fn hex() {
    let source = "0x0123456789ABCDEFabcdef 0x_0123456789ABCDEFabcdef 0x__0123456789ABCDEFabcdef 0x01234_56789ABCDEFabcdef 0x01234__56789ABCDEFabcdef 0x0123456789ABCDEFabcdef_ 0x0123456789ABCDEFabcdef__";
    let tokens = lex(source);

    insta::assert_debug_snapshot!(tokens);
}

#[test]
fn integer() {
    let source = "0 1 1234567890 0_1_2_3_4_5_6_7_8_9_0 0__1__2__3__4__5__6__7__8__9__0 1234567890_ 1234567890__";
    let tokens = lex(source);

    insta::assert_debug_snapshot!(tokens);
}

#[test]
fn float() {
    let source = "0.0 1.0 1234567890.0987654321 0_1_2_3_4_5_6_7_8_9_0.0_9_8_7_6_5_4_3_2_1 0__1__2__3__4__5__6__7__8__9__0.0__9__8__7__6__5__4__3__2__1 1234567890.0987654321_ 1234567890.0987654321__ 1.0e10 1.0E10 1.0e+10 1.0E+10 1.0e-10 1.0E-10 1.0e+10_9_87 1.0E+10__9__8__7 1.0e10_9_87 1.0E10__9__8__7 1e10 1E10 1e+10 1E+10 1e-10 1E-10 1e+10_9_87 1E+10__9__8__7 1e10_9_87 1E10__9__8__7";
    let tokens = lex(source);

    insta::assert_debug_snapshot!(tokens);
}

#[test]
fn identifier() {
    let source = "_ _foo _123 _foo123 foo foo_ foo123 foo123_ foo_foo foo_123 한글";
    let tokens = lex(source);

    insta::assert_debug_snapshot!(tokens);
}

#[test]
fn keywords() {
    let source = "as await break const continue crate else enum false fn for if impl in let loop match mut mod pub return Self self static struct super trait true type use where while yield";
    let tokens = lex(source);

    insta::assert_debug_snapshot!(tokens);
}
