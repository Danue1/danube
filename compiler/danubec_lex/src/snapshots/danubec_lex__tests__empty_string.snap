---
source: compiler/danubec_lex/src/tests.rs
expression: tokens
---
[
    (
        STRING_START,
        "\"",
    ),
    (
        STRING_END,
        "\"",
    ),
    (
        WHITESPACE,
        " ",
    ),
    (
        STRING_START,
        "\"\"\"",
    ),
    (
        STRING_END,
        "\"\"\"",
    ),
]
